Source/Pawnmorphs/Esoteria/Hediffs/Base/Hediff_PM.cs

```diff
@@ -0,0 +1,543 @@
+        /// Called after the hediff is initialized but before it is added to the pawn.
+        /// Any object construction should happen here. 
+        /// </summary>
+        public override void PostMake()
+        {
+            base.PostMake();
+
+            InitializePMComps();
+            for (int index = _pmComps.Count - 1; index >= 0; --index)
+                try
+                {
+                    _pmComps[index]!.CompPostMake();
+                }
+                catch (Exception ex)
+                {
+                    Log.Error("Error in HediffComp_PM.CompPostMake(): " + ex);
+                    _pmComps.RemoveAt(index);
+                }
+        }
+
+        /// <summary>
+        /// Called after the hediff is first added to the pawn.
+        /// Any initialization that requires the pawn to be set should be done here.
+        /// <br />
+        /// Note that this is not called during loading, only when the hediff is initially added.
+        /// </summary>
+        public override void PostAdd(DamageInfo? dInfo)
+        {
+            base.PostAdd(dInfo);
+            foreach (HediffComp_PM pmComp in PMComps)
+                pmComp.CompPostPostAdd(dInfo);
+            PostAddOrLoad();
+        }
+
+        /// <summary>
+        /// Called after the hediff is first added to the pawn, and any time the game is loaded (during the
+        /// <see cref="LoadSaveMode.PostLoadInit" /> phase).
+        /// <br />
+        /// Any initialization that needs to be repeated after every load can be done here.
+        /// </summary>
+        protected virtual void PostAddOrLoad()
+        {
+            HashOffset = pawn?.HashOffset() ?? 0;
+        }
+
+        /// <summary>
+        /// Called after the hediff is removed from the pawn.
+        /// </summary>
+        public override void PostRemoved()
+        {
+            base.PostRemoved();
+            foreach (HediffComp_PM pmComp in PMComps)
+                pmComp.CompPostPostRemoved();
+        }
+
+        //TODO - override Tick() the same way Hediff_StageChanges does
+
+        /// <summary>
+        /// Called after the main Hediff tick.  Comps and PMComps are ticked in this method.
+        /// <br />
+        /// Note that this method is a MAJOR hot path -- it should be optimized as much as possible because it potentially gets
+        /// called tens of thousands of times per second. For this reason, foreach loops should be avoided because the iterator
+        /// object creation is a notable source of slowdown.
+        /// </summary>
+        public override void PostTick()
+        {
+            // We're not calling base.PostTick() here because we're manually replicating it in a slightly more optimized way
+            var severityAdjustment = 0f;
+
+            if (_hasComps && comps != null)
+            {
+                int compCount = comps.Count; // Not caching this genuinely has a noticeable performance impact for hediffs that
```

Comment:
I am amazed that this would have a noticeable effect considering .Count is
`
        // Read-only property describing how many elements are in the List.
        public int Count {
            get {
                Contract.Ensures(Contract.Result<int>() >= 0);
                return _size; 
            }
        }`

https://github.com/microsoft/referencesource/blob/master/mscorlib/system/collections/generic/list.cs#L137

Reply:
The main reason for that is that it's being called several hundred thousand times a second (because it gets called every iteration of the loop, which means once per comp + 1, per hediff, 60 times a second...).  I assume that most of the time spent calling it is actually just all the dynamic dispatching overhead rather than the implementation itself.